src/librustc_target/spec/crt_objects.rs

```diff
@@ -0,0 +1,151 @@
+//! Object code fragments implementing basic runtime facilities and added to the produced binaries.
+//!
+//! Table of CRT objects for popular toolchains.
+//! `crtx` ones are generally distributed with libc and `begin/end` ones with gcc.
+//! See https://dev.gentoo.org/~vapier/crt.txt for some more details.
+//!
+//! | Pre-link CRT objects | glibc                  | musl                   | bionic           | mingw             |
+//! |----------------------|------------------------|------------------------|------------------|-------------------|
+//! | dynamic-nopic-exe    | crt1, crti, crtbegin   | crt1, crti, crtbegin   | crtbegin_dynamic | crt2, crtbegin    |
+//! | dynamic-pic-exe      | Scrt1, crti, crtbeginS | Scrt1, crti, crtbeginS | crtbegin_dynamic | crt2, crtbegin    |
+//! | static-nopic-exe     | crt1, crti, crtbeginT  | crt1, crti, crtbegin   | crtbegin_static  | crt2, crtbegin    |
+//! | static-pic-exe       | rcrt1, crti, crtbeginS | rcrt1, crti, crtbeginS | crtbegin_dynamic | crt2, crtbegin    |
+//! | dynamic-dylib        | crti, crtbeginS        | crti, crtbeginS        | crtbegin_so      | dllcrt2, crtbegin |
+//! | static-dylib (gcc)   | crti, crtbeginT        | crti, crtbeginS        | crtbegin_so      | dllcrt2, crtbegin |
+//! | static-dylib (clang) | crti, crtbeginT        | N/A                    | crtbegin_static  | dllcrt2, crtbegin |
+//!
+//! | Post-link CRT objects | glibc         | musl          | bionic         | mingw  |
+//! |-----------------------|---------------|---------------|----------------|--------|
+//! | dynamic-nopic-exe     | crtend, crtn  | crtend, crtn  | crtend_android | crtend |
+//! | dynamic-pic-exe       | crtendS, crtn | crtendS, crtn | crtend_android | crtend |
+//! | static-nopic-exe      | crtend, crtn  | crtend, crtn  | crtend_android | crtend |
+//! | static-pic-exe        | crtendS, crtn | crtendS, crtn | crtend_android | crtend |
+//! | dynamic-dylib         | crtendS, crtn | crtendS, crtn | crtend_so      | crtend |
+//! | static-dylib (gcc)    | crtend, crtn  | crtendS, crtn | crtend_so      | crtend |
+//! | static-dylib (clang)  | crtendS, crtn | N/A           | crtend_so      | crtend |
+//!
+//! If rustc adds the objects itself rather than relies on what gcc pulls,
+//! then it normally adds only the libc's objects, but not gcc's.
```

Comment:
https://github.com/rust-lang/rust/pull/40113#issuecomment-311049710 reports that we need to pull the gcc's objects (crtbegin*, crtend*) as well if we want to link with static libraries written in C++.